File: day16/floydwarshall.go
package main

type (
	PairwiseDistances[T comparable] map[T]map[T]int

	Edge[T any] struct {
		A, B T
		C    int
	}
)

func (m PairwiseDistances[T]) EdgeCost(a, b T) (int, bool) {
	d, ok := m[a]
	if !ok {
		return 0, false
	}
	k, ok := d[b]
	if !ok {
		return 0, false
	}
	return k, true
}

func FloydWarshall[T comparable](nodes []T, edges []Edge[T]) map[T]map[T]int {
	dist := PairwiseDistances[T]{}
	for _, i := range edges {
		if _, ok := dist[i.A]; !ok {
			dist[i.A] = map[T]int{}
		}
		dist[i.A][i.B] = i.C
	}
	for _, i := range nodes {
		if _, ok := dist[i]; !ok {
			dist[i] = map[T]int{}
		}
		dist[i][i] = 0
	}
	for _, k := range nodes {
		for _, i := range nodes {
			for _, j := range nodes {
				cik, ok := dist.EdgeCost(i, k)
				if !ok {
					continue
				}
				ckj, ok := dist.EdgeCost(k, j)
				if !ok {
					continue
				}
				ck := cik + ckj
				if cij, ok := dist.EdgeCost(i, j); !ok || cij > ck {
					dist[i][j] = ck
				}
			}
		}
	}
	return dist
}

File: day16/src/floydwarshall.rs
use std::collections::HashMap;
use std::hash::Hash;

pub struct Edge<T> {
    pub a: T,
    pub b: T,
    pub c: i32,
}

pub struct PairwiseDistances<T> {
    arr: HashMap<T, HashMap<T, i32>>,
}

impl<T> PairwiseDistances<T>
where
    T: Eq + Hash,
{
    fn new() -> Self {
        Self {
            arr: HashMap::new(),
        }
    }

    pub fn edge_cost(&self, a: &T, b: &T) -> Option<i32> {
        self.arr
            .get(a)
            .and_then(|m| m.get(b))
            .and_then(|&k| Some(k))
    }

    fn set(&mut self, a: T, b: T, c: i32) {
        self.arr.entry(a).or_insert(HashMap::new()).insert(b, c);
    }
}

pub fn compute<T>(nodes: &[T], edges: &[Edge<T>]) -> PairwiseDistances<T>
where
    T: Clone + Eq + Hash,
{
    let mut dist = PairwiseDistances::new();
    for i in edges {
        dist.set(i.a.clone(), i.b.clone(), i.c);
    }
    for i in nodes {
        dist.set(i.clone(), i.clone(), 0);
    }
    for k in nodes {
        for i in nodes {
            for j in nodes {
                if let Some(ck) = dist
                    .edge_cost(&i, &k)
                    .and_then(|cik| dist.edge_cost(&k, &j).and_then(|ckj| Some(cik + ckj)))
                {
                    if match dist.edge_cost(&i, &j) {
                        Some(cij) => cij > ck,
                        None => true,
                    } {
                        dist.set(i.clone(), j.clone(), ck);
                    }
                }
            }
        }
    }
    dist
}
